File: internal/apperr/apperr.go
// Package apperr defines the stable error model for sph: machine-readable
// error codes, a processing stage, safe human messages and stable process
// exit codes.
//
// Message MUST stay safe to print: no cookies, session headers, generalToken,
// signed media URLs or raw upstream bodies. The wrapped Err is never rendered.
package apperr

import (
	"context"
	"errors"
	"fmt"
)

// Code is a stable machine-readable error identifier.
type Code string

const (
	InternalError          Code = "INTERNAL_ERROR"
	InvalidArgument        Code = "INVALID_ARGUMENT"
	AuthRequired           Code = "AUTH_REQUIRED"
	InvalidCredentials     Code = "INVALID_CREDENTIALS"
	NetworkError           Code = "NETWORK_ERROR"
	Timeout                Code = "TIMEOUT"
	UpstreamError          Code = "UPSTREAM_ERROR"
	AccessDenied           Code = "ACCESS_DENIED"
	RateLimited            Code = "RATE_LIMITED"
	VideoUnavailable       Code = "VIDEO_UNAVAILABLE"
	NoMedia                Code = "NO_MEDIA"
	UnsupportedMedia       Code = "UNSUPPORTED_MEDIA"
	DownloadFailed         Code = "DOWNLOAD_FAILED"
	DownloadTooLarge       Code = "DOWNLOAD_TOO_LARGE"
	FileExists             Code = "FILE_EXISTS"
	IOError                Code = "IO_ERROR"
	VerifyFailed           Code = "VERIFY_FAILED"
	SchemaChanged          Code = "SCHEMA_CHANGED"
	LoginDependencyMissing Code = "LOGIN_DEPENDENCY_MISSING"
	LoginBrowserFailed     Code = "LOGIN_BROWSER_FAILED"
	InteractiveRequired    Code = "INTERACTIVE_REQUIRED"
	LoginCaptureFailed     Code = "LOGIN_CAPTURE_FAILED"
	AuthBusy               Code = "AUTH_BUSY"
	Cancelled              Code = "CANCELLED"
)

// Stage identifies where in the pipeline an error occurred.
type Stage string

const (
	StageArguments    Stage = "arguments"
	StageCredentials  Stage = "credentials"
	StageLoginDeps    Stage = "login_dependencies"
	StageLoginBrowser Stage = "login_browser"
	StageLoginCapture Stage = "login_capture"
	StageLoginCommit  Stage = "login_commit"
	StageParseShare   Stage = "parse_share"
	StageFetchFeed    Stage = "fetch_feed"
	StageSelectMedia  Stage = "select_media"
	StageDownload     Stage = "download"
	StageVerify       Stage = "verify"
	StageCommit       Stage = "commit"
)

// exitCodes maps every code to its process exit code.
var exitCodes = map[Code]int{
	InternalError:          1,
	InvalidArgument:        2,
	AuthRequired:           3,
	InvalidCredentials:     3,
	NetworkError:           4,
	Timeout:                4,
	UpstreamError:          5,
	AccessDenied:           5,
	RateLimited:            5,
	VideoUnavailable:       6,
	NoMedia:                6,
	UnsupportedMedia:       7,
	DownloadFailed:         8,
	DownloadTooLarge:       8,
	FileExists:             9,
	IOError:                9,
	VerifyFailed:           10,
	SchemaChanged:          11,
	LoginDependencyMissing: 12,
	LoginBrowserFailed:     12,
	InteractiveRequired:    13,
	LoginCaptureFailed:     14,
	AuthBusy:               14,
	Cancelled:              130,
}

// Error is the error type carried between internal packages.
type Error struct {
	Code      Code
	Stage     Stage
	Message   string
	Retryable bool
	Err       error // underlying cause; never printed
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s (%s): %s", e.Code, e.Stage, e.Message)
}

func (e *Error) Unwrap() error { return e.Err }

// New builds an *Error without an underlying cause.
func New(code Code, stage Stage, msg string) *Error {
	return &Error{Code: code, Stage: stage, Message: msg}
}

// Newf is New with formatting.
func Newf(code Code, stage Stage, format string, args ...any) *Error {
	return &Error{Code: code, Stage: stage, Message: fmt.Sprintf(format, args...)}
}

// Wrap attaches an underlying cause that is kept for diagnostics but never
// rendered to the user.
func Wrap(err error, code Code, stage Stage, msg string) *Error {
	return &Error{Code: code, Stage: stage, Message: msg, Err: err}
}

// From coerces any error into an *Error. Known context errors are classified;
// anything else becomes INTERNAL_ERROR unless it already is an *Error.
func From(err error) *Error {
	if err == nil {
		return nil
	}
	var ae *Error
	if errors.As(err, &ae) {
		return ae
	}
	switch {
	case errors.Is(err, context.Canceled):
		return Wrap(err, Cancelled, StageArguments, "操作已取消")
	case errors.Is(err, context.DeadlineExceeded):
		return Wrap(err, Timeout, StageArguments, "操作超时")
	}
	return Wrap(err, InternalError, StageArguments, "未预期的内部错误")
}

// ExitCode maps err to the stable process exit code; nil maps to 0.
func ExitCode(err error) int {
	if err == nil {
		return 0
	}
	if code, ok := exitCodes[From(err).Code]; ok {
		return code
	}
	return 1
}

// CodeOf returns the apperr code of err, or "" when err is nil.
func CodeOf(err error) Code {
	if err == nil {
		return ""
	}
	return From(err).Code
}

// StageOf returns the stage of err, or "" when err is nil.
func StageOf(err error) Stage {
	if err == nil {
		return ""
	}
	return From(err).Stage
}

File: src/apperr.rs
//! 稳定的错误模型：机器可读错误码、处理阶段、安全的人类可读消息和稳定的进程退出码。
//!
//! 消息必须保持可打印的安全性：不得包含 Cookie、会话头、generalToken、带签名的媒体
//! URL 或上游原始响应体。包装的根本原因永不渲染。

/// 机器可读错误码。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Code {
    InternalError,
    InvalidArgument,
    AuthRequired,
    InvalidCredentials,
    NetworkError,
    Timeout,
    UpstreamError,
    AccessDenied,
    RateLimited,
    VideoUnavailable,
    NoMedia,
    UnsupportedMedia,
    DownloadFailed,
    DownloadTooLarge,
    FileExists,
    IOError,
    VerifyFailed,
    SchemaChanged,
    LoginDependencyMissing,
    LoginBrowserFailed,
    InteractiveRequired,
    LoginCaptureFailed,
    AuthBusy,
    Cancelled,
    // v2 新增（PRD §8）
    SessionExpired,
    PublishRejected,
    RecoveryFailed,
    ScheduleInvalid,
}

impl Code {
    pub fn as_str(self) -> &'static str {
        match self {
            Code::InternalError => "INTERNAL_ERROR",
            Code::InvalidArgument => "INVALID_ARGUMENT",
            Code::AuthRequired => "AUTH_REQUIRED",
            Code::InvalidCredentials => "INVALID_CREDENTIALS",
            Code::NetworkError => "NETWORK_ERROR",
            Code::Timeout => "TIMEOUT",
            Code::UpstreamError => "UPSTREAM_ERROR",
            Code::AccessDenied => "ACCESS_DENIED",
            Code::RateLimited => "RATE_LIMITED",
            Code::VideoUnavailable => "VIDEO_UNAVAILABLE",
            Code::NoMedia => "NO_MEDIA",
            Code::UnsupportedMedia => "UNSUPPORTED_MEDIA",
            Code::DownloadFailed => "DOWNLOAD_FAILED",
            Code::DownloadTooLarge => "DOWNLOAD_TOO_LARGE",
            Code::FileExists => "FILE_EXISTS",
            Code::IOError => "IO_ERROR",
            Code::VerifyFailed => "VERIFY_FAILED",
            Code::SchemaChanged => "SCHEMA_CHANGED",
            Code::LoginDependencyMissing => "LOGIN_DEPENDENCY_MISSING",
            Code::LoginBrowserFailed => "LOGIN_BROWSER_FAILED",
            Code::InteractiveRequired => "INTERACTIVE_REQUIRED",
            Code::LoginCaptureFailed => "LOGIN_CAPTURE_FAILED",
            Code::AuthBusy => "AUTH_BUSY",
            Code::SessionExpired => "SESSION_EXPIRED",
            Code::PublishRejected => "PUBLISH_REJECTED",
            Code::RecoveryFailed => "RECOVERY_FAILED",
            Code::ScheduleInvalid => "SCHEDULE_INVALID",
            Code::Cancelled => "CANCELLED",
        }
    }

    /// 稳定的进程退出码。
    pub fn exit_code(self) -> i32 {
        match self {
            Code::InternalError => 1,
            Code::InvalidArgument => 2,
            Code::AuthRequired | Code::InvalidCredentials => 3,
            Code::NetworkError | Code::Timeout => 4,
            Code::UpstreamError | Code::AccessDenied | Code::RateLimited => 5,
            Code::VideoUnavailable | Code::NoMedia => 6,
            Code::UnsupportedMedia => 7,
            Code::DownloadFailed | Code::DownloadTooLarge => 8,
            Code::FileExists | Code::IOError => 9,
            Code::VerifyFailed => 10,
            Code::SchemaChanged => 11,
            Code::LoginDependencyMissing | Code::LoginBrowserFailed => 12,
            Code::InteractiveRequired => 13,
            Code::LoginCaptureFailed | Code::AuthBusy => 14,
            Code::SessionExpired => 15,
            Code::PublishRejected => 16,
            Code::RecoveryFailed => 17,
            Code::ScheduleInvalid => 18,
            Code::Cancelled => 130,
        }
    }
}

/// 出错阶段。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Stage {
    Arguments,
    Credentials,
    LoginDeps,
    LoginBrowser,
    LoginCapture,
    LoginCommit,
    ParseShare,
    FetchFeed,
    SelectMedia,
    Download,
    Verify,
    Commit,
    // v2 发布流水线
    SessionLoad,
    Navigate,
    Upload,
    Metadata,
    Cover,
    Declaration,
    Schedule,
    Submit,
    Recovery,
}

impl Stage {
    pub fn as_str(self) -> &'static str {
        match self {
            Stage::Arguments => "arguments",
            Stage::Credentials => "credentials",
            Stage::LoginDeps => "login_dependencies",
            Stage::LoginBrowser => "login_browser",
            Stage::LoginCapture => "login_capture",
            Stage::LoginCommit => "login_commit",
            Stage::ParseShare => "parse_share",
            Stage::FetchFeed => "fetch_feed",
            Stage::SelectMedia => "select_media",
            Stage::Download => "download",
            Stage::Verify => "verify",
            Stage::Commit => "commit",
            Stage::SessionLoad => "session_load",
            Stage::Navigate => "navigate",
            Stage::Upload => "upload",
            Stage::Metadata => "metadata",
            Stage::Cover => "cover",
            Stage::Declaration => "declaration",
            Stage::Schedule => "schedule",
            Stage::Submit => "submit",
            Stage::Recovery => "recovery",
        }
    }
}

/// 内部包之间传递的错误类型。
#[derive(Debug, Clone)]
pub struct AppError {
    pub code: Code,
    pub stage: Stage,
    pub message: String,
    pub retryable: bool,
}

impl AppError {
    pub fn new(code: Code, stage: Stage, msg: impl Into<String>) -> AppError {
        AppError {
            code,
            stage,
            message: msg.into(),
            retryable: false,
        }
    }

    pub fn retryable(code: Code, stage: Stage, msg: impl Into<String>) -> AppError {
        AppError {
            code,
            stage,
            message: msg.into(),
            retryable: true,
        }
    }

    pub fn fmt(code: Code, stage: Stage, args: std::fmt::Arguments) -> AppError {
        AppError::new(code, stage, format!("{args}"))
    }
}

impl std::fmt::Display for AppError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} ({}): {}",
            self.code.as_str(),
            self.stage.as_str(),
            self.message
        )
    }
}

impl std::error::Error for AppError {}

/// 结果别名。
pub type Result<T> = std::result::Result<T, AppError>;

/// 将任意错误归入错误模型；已是 AppError 的原样返回。
pub fn from(err: impl Into<BoxError>) -> AppError {
    let err = err.into();
    match err.downcast::<AppError>() {
        Ok(ae) => *ae,
        Err(other) => {
            let msg = format!("{other}");
            if msg.contains("operation timed out") {
                AppError::new(Code::Timeout, Stage::Arguments, "操作超时")
            } else {
                AppError::new(Code::InternalError, Stage::Arguments, "未预期的内部错误")
            }
        }
    }
}

pub type BoxError = Box<dyn std::error::Error + Send + Sync>;

/// 方便的 ? 转换：任意实现 Error 的类型经此包装为内部错误，除非它本就是 AppError。
pub fn wrap_internal(stage: Stage, msg: impl Into<String>) -> impl FnOnce(BoxError) -> AppError {
    let msg = msg.into();
    move |e| {
        let _ = e;
        AppError::new(Code::InternalError, stage, msg)
    }
}
